### PR TITLE
Integration with ArviZ

### PR DESCRIPTION
**Pull request by rlouf**
ArviZ is the best package for the exploratory analysis of models, it makes sense to provide a seamless integration with the library. I chose to implement the `Trace` object as a subclass of Arviz's `InferenceData` so MCX traces can be directly used in ArviZ.
There is a loss of information in that translation since the sampler returns pretty much all there is to know about the process, but we can get that information (say for debugging) using iterative sampling.

- [x] Transform the chain produced by the sampler w/ progress bar so it has the same format as the chain produced by the accelerated `lax.scan` loop;
- [x] Modify the HMC evaluator class for the new format (should be much faster);
- [x] Check `io_numpyro.py`;
- [x] First mvp constructing an `InferenceData` instance with chain positions. [Test plotting](https://twitter.com/remilouf/status/1312132609435037696?s=20).
- [x] Add as much relevant info as reasonable in the chain; keep others (like the proposed step at each step) for the output of iterative sampling;
- [x] Add `log_likelihood` as a `@property` (it requires extra computation).
- [x] Check that trace_1 + trace_2, warmup + trace, prior + trace, etc. work as expected
- [x] Warmup chains and info
- [x] Refactor `HMCInfo`
- [x] Add `append` method to add a single sample to the trace. Should be fast.
- [x] Test with every diagnostic & plot available in ArviZ
- [x] benchmark chain stacking methods
- [x] Comprehensive documentation
- [x] Claim victory and merge

### CODE REVIEW

**Comment by OriolAbril** [comment]
I'll try to keep an eye and help wherever possible, I should have time to review and answer questions about InferenceData.

**Comment by rlouf** [comment]
Thanks! I'm doing a lot of code reading and trial/error at the moment, I'll ping you when I have something tangible!

File: mcx/trace.py [hunk 1]
```diff
@@ -1,31 +1,44 @@
...
-            "posterior_predictive": self.posterior_predictive,
-            "sample_stats": self.sample_stats,
-            "log_likelihood": self.log_likelihood,
+        samples_dataset = dict_to_dataset(
+            data=posterior["samples"], library=mcx, coords=coords, dims=dims
+        )
+
+        sample_stats_dict = {
+            "lp": posterior["potential_energy"],
+            "mean_tree_accept": posterior["acceptance_probability"],
```
Comment (OriolAbril) [inline mcx/trace.py:33]:
```suggestion
            "acceptance_rate": posterior["acceptance_probability"],
```

We are working on a naming convention for sample stats. Right now only lp and diverging are enforced, the other stats have different names depending on the sampling library. 

Currently ArviZ only uses `diverging` and `energy` in its plots, but with time more quantities will probably be used, probably even more now with [arviz_dashboard](https://github.com/arviz-devs/arviz_dashboard).

Comment (rlouf) [reply mcx/trace.py:33]: Perfect timing! I was going to ask you about this since many names are NUTS-specific, and I have only implemented HMC so far.

**Comment by rlouf** [comment]
@OriolAbril Several of the metrics in `sample_stats` are noticeably slowing down inference for large sample sizes. Some I can probably improve (I have to transpose a couple of arrays to fit the format), but for others like `log_probability` I need to perform more calculations. 

It can be problematic when people are first iterating in a model and only need the `plot_trace`. Why slow them down for metrics they don't need?

So I was wondering if you thought overriding some of the class' attributes with a `@property` would be reasonable in this case?

**Comment by OriolAbril** [comment]
Not sure I understand, you mean so that `lp` and other sample stats are available but only computed whenever needed?

I think we have not considered it (could be wrong though). 

What we currently have for the `log_likelihood` (it is useful for loo/waic but otherwise requires extra memory and extra computing) group is an argument to `from_pymc3`, `from_pyro`... so that users can choose whether or not the group is to be included in the resulting InferenceData. The same could be done for sample_stats and its variables.

**Comment by rlouf** [comment]
>Not sure I understand, you mean so that `lp` and other sample stats are
>available but only computed whenever needed?

Yes, I could do that by defining a `sample_stats` method in my `Trace`
class that overrides the corresponding attribute in `InferenceData`.
Do I make sense?

>I think we have not considered it (could be wrong though).

I don't necessarily think you would need to do that on your end.

>What we currently have for the `log_likelihood` (it is useful for
>loo/waic but otherwise requires extra memory and extra computing) group
>is an argument to `from_pymc3`, `from_pyro`... so that users can choose
>whether or not the group is to be included in the resulting
>InferenceData. The same could be done for sample_stats and its
>variables.

I did not see that, I will check it out!

**Comment by rlouf** [comment]
Nevermind, I made some simple benchmarks and it looks like the running time is only marginally affected by all the conversions. I believe I also found a way to get the `log_likelihood` for free during sampling by making a small change to my compiled logpdf.

**Comment by rlouf** [comment]
I can confirm that all plots / stats that take an `InferenceData` instance as an input and do not require `log_likelihood` work fine.

**Comment by rlouf** [comment]
Computation of `log_likelihood` working, moving on to include the warmup trace.

**Comment by rlouf** [comment]
Concatenation works.

**Comment by rlouf** [comment]
Yay! Just need to improve documentation a little!

**Comment by rlouf** [comment]
Found a performance issue linked to the fact that `lax.scan` and the `for` loop do not output the chain in the same format. Reported in #41. I found a fix, but will implement in another PR.

**Pull request merged**